kerastuner/engine/base_tuner.py

```diff
@@ -0,0 +1,232 @@
+        self._save_trial(trial)
+        self.save()
+
+    def on_search_end(self):
+        if self.logger:
+            self.logger.exit()
+
+    def get_best_models(self, num_models=1):
+        """Returns the best model(s), as determined by the tuner's objective.
+
+        Args:
+            num_models (int, optional): Number of best models to return.
+                Models will be returned in sorted order. Defaults to 1.
+
+        Returns:
+            List of trained model instances.
+        """
+        raise NotImplementedError
+
+    def search_space_summary(self, extended=False):
+        """Print search space summary.
+
+        Args:
+            extended: Bool, optional. Display extended summary.
+                Defaults to False.
+        """
+        display.section('Search space summary')
+        hp = self.oracle.get_space()
+        display.display_setting(
+            'Default search space size: %d' % len(hp.space))
+        for p in hp.space:
+            config = p.get_config()
+            name = config.pop('name')
+            display.subsection('%s (%s)' % (name, p.__class__.__name__))
+            display.display_settings(config)
+
+    def results_summary(self, num_trials=10):
+        """Display tuning results summary.
+
+        Args:
+            num_trials (int, optional): Number of trials to display.
+                Defaults to 10.
+            sort_metric (str, optional): Sorting metric, when not specified
+                sort models by objective value. Defaults to None.
+        """
+        display.section('Results summary')
+        display.display_setting('Results in %s' % self._get_project_dir())
+        best_trials = self.oracle.get_best_trials(num_trials)
+        display.display_setting('Showing %d best trials' % num_trials)
+        for trial in best_trials:
+            display.display_setting(
+                'Objective: {} Score: {}'.format(
+                    self.oracle.objective, trial.score))
+
+    @property
+    def remaining_trials(self):
+        return self.oracle.remaining_trials()
+
+    def get_state(self):
+        return {}
+
+    def set_state(self, state):
+        pass
+
+    def save(self):
+        self.oracle.save(self._get_oracle_fname())
+        super(BaseTuner, self).save(self._get_tuner_fname())
+
+    def reload(self):
+        self.oracle.reload(self._get_oracle_fname())
+        super(BaseTuner, self).reload(self._get_tuner_fname())
+
+    def _save_trial(self, trial):
+        # Write trial status to trial directory
+        trial.save(self._get_trial_fname(trial))
+        # Send status to Logger
+        if self.logger:
+            self.logger.report_trial_state(trial.trial_id, trial.get_state())
+
+    def _get_project_dir(self):
+        dirname = os.path.join(
+            self.directory,
+            self.project_name)
+        tf_utils.create_directory(dirname)
+        return dirname
+
+    def _get_trial_dir(self, trial):
```

Comment:
Shouldn't this method use `trial_id` as input? Also, this would be useful method to make public, since both subclassers of the base tuner and users of any tuner may need to use it. Same for project dir.

Reply:
Done